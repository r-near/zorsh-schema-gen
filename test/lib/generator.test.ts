import { readFileSync } from "node:fs"
import { join } from "node:path"
import { describe, expect, it } from "vitest"
import { BorshSchemaContainerSchema, ZorshGenerator } from "../../src/lib/index.js"

describe("ZorshGenerator", () => {
  it("should generate valid schema from complex example", () => {
    // Read the schema.bin
    const schemaPath = join(__dirname, "..", "fixtures", "complex_schema.bin")
    const data = readFileSync(schemaPath)

    // Parse it
    const container = BorshSchemaContainerSchema.deserialize(data)

    // Generate code
    const generator = new ZorshGenerator()
    const code = generator.generate(container)

    expect(code).toMatchSnapshot()

    // Basic validation
    expect(code).toContain('import { b } from "@zorsh/zorsh"')

    // Check for expected type definitions
    expect(code).toContain("export const LocationSchema = b.struct({")
    expect(code).toContain("export const ItemRaritySchema = b.enum({")
    expect(code).toContain("export const StatsSchema = b.struct({")
    expect(code).toContain("export const PlayerCharacterSchema = b.struct({")

    // Check specific field types
    expect(code).toContain("lat: b.f64()")
    expect(code).toContain("lng: b.f64()")
    expect(code).toContain("altitude: b.option(b.u32())")

    // Check enum variants
    expect(code).toContain("Common: b.unit()")
    expect(code).toContain("Legendary: b.unit()")

    // Check collection types
    expect(code).toContain("b.vec(")
    expect(code).toContain("b.hashMap(")
    expect(code).toContain("b.hashSet(")

    // Check that generated code exports types
    expect(code).toContain("export type Location = b.infer<typeof LocationSchema>;")
  })
})
