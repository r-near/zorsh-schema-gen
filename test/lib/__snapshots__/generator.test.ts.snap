// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ZorshGenerator > should generate valid schema from complex example 1`] = `
"import { b } from "@zorsh/zorsh";


export const AchievementSchema = b.struct({
  id: b.string(),
  name: b.string(),
  description: b.string(),
  reward_points: b.u32(),
  completed_at: b.option(b.u64()),
  prerequisites: b.vec(b.string())
});
export type Achievement = b.infer<typeof AchievementSchema>;

export const StatsSchema = b.struct({
  strength: b.u16(),
  dexterity: b.u16(),
  intelligence: b.u16(),
  health: b.u32(),
  mana: b.u32()
});
export type Stats = b.infer<typeof StatsSchema>;

export const CharacterClassSchema = b.enum({
  Warrior: StatsSchema,
  Mage: b.struct({
    base_stats: StatsSchema,
    spells_known: b.vec(b.string()),
    mana_regen: b.f32()
  }),
  Rogue: b.struct({
    base_stats: StatsSchema,
    stealth_level: b.u8(),
    critical_chance: b.f32()
  })
});
export type CharacterClass = b.infer<typeof CharacterClassSchema>;

export const GuildMembershipSchema = b.struct({
  guild_id: b.string(),
  guild_name: b.string(),
  joined_at: b.u64(),
  rank: b.string(),
  permissions: b.hashSet(b.string())
});
export type GuildMembership = b.infer<typeof GuildMembershipSchema>;

export const ItemRaritySchema = b.enum({
  Common: b.unit(),
  Uncommon: b.unit(),
  Rare: b.unit(),
  Epic: b.unit(),
  Legendary: b.unit()
});
export type ItemRarity = b.infer<typeof ItemRaritySchema>;

export const ItemEffectSchema = b.enum({
  Damage: b.u32(),
  Heal: b.struct({
    amount: b.u32(),
    duration: b.u16()
  }),
  Status: b.struct({
    effect_type: b.string(),
    power: b.u16(),
    duration: b.u32()
  }),
  None: b.unit()
});
export type ItemEffect = b.infer<typeof ItemEffectSchema>;

export const ItemSchema = b.struct({
  id: b.string(),
  name: b.string(),
  description: b.string(),
  rarity: ItemRaritySchema,
  level_requirement: b.u8(),
  effects: b.vec(ItemEffectSchema),
  stats: b.option(StatsSchema),
  metadata: b.hashMap(b.string(), b.string())
});
export type Item = b.infer<typeof ItemSchema>;

export const LocationSchema = b.struct({
  lat: b.f64(),
  lng: b.f64(),
  altitude: b.option(b.u32())
});
export type Location = b.infer<typeof LocationSchema>;

export const QuestProgressSchema = b.struct({
  quest_id: b.string(),
  started_at: b.u64(),
  steps_completed: b.vec(b.u32()),
  current_step: b.u32(),
  collected_items: b.hashMap(b.string(), b.u32())
});
export type QuestProgress = b.infer<typeof QuestProgressSchema>;

export const TradeSchema = b.struct({
  id: b.string(),
  from_player: b.string(),
  to_player: b.string(),
  items: b.vec(ItemSchema),
  gold_amount: b.u64(),
  status: b.string(),
  created_at: b.u64(),
  completed_at: b.option(b.u64())
});
export type Trade = b.infer<typeof TradeSchema>;

export const PlayerCharacterSchema = b.struct({
  id: b.string(),
  name: b.string(),
  created_at: b.u64(),
  last_login: b.u64(),
  character_class: CharacterClassSchema,
  level: b.u32(),
  experience: b.u64(),
  location: LocationSchema,
  inventory: b.vec(ItemSchema),
  equipped_items: b.hashMap(b.string(), ItemSchema),
  achievements: b.vec(AchievementSchema),
  completed_quests: b.hashSet(b.string()),
  active_quests: b.vec(QuestProgressSchema),
  guild: b.option(GuildMembershipSchema),
  current_trades: b.vec(TradeSchema),
  friends: b.hashSet(b.string()),
  blocked_players: b.hashSet(b.string()),
  settings: b.hashMap(b.string(), b.string()),
  last_deaths: b.vec(b.tuple([b.u64(), LocationSchema, b.string()])),
  skill_levels: b.hashMap(b.string(), b.tuple([b.u16(), b.f32()]))
});
export type PlayerCharacter = b.infer<typeof PlayerCharacterSchema>;"
`;
